src/stream/mod.rs

```diff
@@ -2,3 +2,49 @@
 
 pub mod log;
 pub mod stdin;
+
+use crate::{Captures, Needle, Error};
+use std::io;
+
+#[cfg(not(feature = "async"))]
+use std::io::{Write, Read, BufRead};
+
+#[cfg(feature = "async")]
+use futures_lite::{AsyncRead, AsyncWrite, AsyncBufRead};
+
+/// Trait for types that can read and write to child programs.
+#[cfg(not(feature = "async"))]
+pub trait StreamSink: Write + Read + BufRead {
+    /// Send a buffer to the child program.
+    fn send<B: AsRef<[u8]>>(&mut self, buf: B) -> io::Result<()>;
+    
+    /// Send a line to the child program.
+    fn send_line(&mut self, text: &str) -> io::Result<()>;
+    
+    /// Expect output from the child program.
+    fn expect<N>(
+        &mut self,
+        needle: N,
+    ) -> Result<Captures, Error>
+    where
+        N: Needle;
+}
+
+/// Trait for types that can read and write to child programs.
+#[cfg(feature = "async")]
+#[async_trait::async_trait(?Send)]
+pub trait StreamSink: AsyncRead + AsyncWrite + AsyncBufRead + Unpin {
```

Comment:
Wouldn't it be better to have `StreamSink` and `StreamSinkAsync`?
*Not sure about name though.

Reply:
As an example ;
so even in `async` the implementation could could both `StreamSink+StreamSinkAsync` as a general case just with some lock.

Reply:
I think it's better to use the same name regardless of sync/async. Using sync and async code from a library is typically mutually exclusive (and I think mixing should be discouraged), so if I start writing my program sync and decide to migrate to `async` it should be transparent for type names, all I should need to do is add `async` and `.await` in the appropriate places.

Reply:
Not sure some, but maybe